subprojects/robotpy-wpilib/wpilib/iterativerobotpy.py

```diff
@@ -0,0 +1,237 @@
+            self._m_watchdog.addEpoch("DisabledInit()")
+        elif mode == IterativeRobotMode.kAutonomous:
+            self.autonomousInit()
+            self._m_watchdog.addEpoch("AutonomousInit()")
+        elif mode == IterativeRobotMode.kTeleop:
+            self.teleopInit()
+            self._m_watchdog.addEpoch("TeleopInit()")
+        elif mode == IterativeRobotMode.kTest:
+            if self._m_lwEnabledInTest:
+                wpilib.LiveWindow.setEnabled(True)
+                # todo Shuffleboard::EnableActuatorWidgets()
+            self.testInit()
+            self._m_watchdog.addEpoch("TestInit()")
+        self._m_lastMode = mode
+
+        # Call the appropriate function depending upon the current robot mode
+        if mode == IterativeRobotMode.kDisabled:
+            hal.observeUserProgramDisabled()
+            self.disabledPeriodic()
+            self._m_watchdog.addEpoch("DisabledPeriodic()")
+        elif mode == IterativeRobotMode.kAutonomous:
+            hal.observeUserProgramAutonomous()
+            self.autonomousPeriodic()
+            self._m_watchdog.addEpoch("AutonomousPeriodic()")
+        elif mode == IterativeRobotMode.kTeleop:
+            hal.observeUserProgramTeleop()
+            self.teleopPeriodic()
+            self._m_watchdog.addEpoch("TeleopPeriodic()")
+        elif mode == IterativeRobotMode.kTest:
+            hal.observeUserProgramTest()
+            self.testPeriodic()
+            self._m_watchdog.addEpoch("TestPeriodic()")
+
+        self.robotPeriodic()
+        self._m_watchdog.addEpoch("RobotPeriodic()")
+        #
+        wpilib.SmartDashboard.updateValues()
+        self._m_watchdog.addEpoch("SmartDashboard::UpdateValues()")
+        wpilib.LiveWindow.updateValues()
+        self._m_watchdog.addEpoch("LiveWindow::UpdateValues()")
+        # todo Shuffleboard::Update()
+        self._m_watchdog.addEpoch("Shuffleboard::Update()")
+        if self.isSimulation():
+            hal.simPeriodicBefore()
+            self.simulationPeriodic()
+            hal.simPeriodicAfter()
+            self._m_watchdog.addEpoch("SimulationPeriodic()")
+
+        self._m_watchdog.Disable()
+
+        # // Flush NetworkTables
+        if self._m_ntFlushEnabled:
+            ntcore.NetworkTableInstance.getDefault().flushLocal()
+
+        # Warn on loop time overruns
+        if self._m_watchdog.IsExpired():
+            self._m_watchdog.PrintEpochs()
+
+    def printLoopOverrunMessages(self):
+        # todo ask about this
+        # cpp has this as a error, java as a warning, is this the right way to call?
+        # void IterativeRobotBase::PrintLoopOverrunMessage() {
+        #     FRC_ReportError(err::Error, "Loop time of {:.6f}s overrun", m_period.value())
+        # }
+        # private void printLoopOverrunMessage() {
+        #     DriverStation.reportWarning("Loop time of " + m_period + "s overrun\n", false);
+        # }
+        wpilib._impl.report_error.reportWarning(
```

Comment:
Yes, this is correct.

I would do at the top of this file `from wpilib import reportWarning`. Every time there's a `.` in python it's another dictionary lookup, and the rio is really slow.